### PR TITLE
Create profiles with names mutations

### LINKED ISSUE #1395: Move names to profiles
This feature moves a user's `name` field from being a db field attached to `users` (and being different per each circle) to having one single name per address.

The following changes are required to implement this:

Phase 1 - DB / App Changes
* add column `name` on `profiles` table.
* Use `profiles.name` instead of `users.name` across the site -- fall back to `users.name` only when `profiles.name` is not set
* If an address has only one name (either in only one circle, or has the same name across all circles) automatically copy this user's name into their new `profile.name`
* If an address has multiple names: do not set any profile name for them, and prompt them to complete their profile info on next login.

Phase 2 - Updated Username Functionality
* Remove ability for circle admin to change a users name (remove from settings page, leave for add member as initial value)
* Add "change name" option to profile edit page

Phase 3 - Rollout and comms
* [Optional] think about notifying users with different names about this new options

Phase 4 - clean up:
* drop column `name` on `users` table once no longer in use and rollout was successful

----

### Detailed to-do sequence (work in progress)

- [x] Resolve conflicts (users with same name and different addresses)
  - [x] Sample users
  - [x] Actual users
    - there are over 800 name conflicts: ```select name, count(distinct address) ac from users group by name having count(distinct address) > 1 order by ac desc```
    - [x] create a migration for this: just append a number to each name
- [ ] Set `profiles.name` site-wide
  - due to the above, once this query is changed, the only profiles without names should now be ones with the same address and different names
  - we don't want to automatically pick one of these names for the address because it could cause information leakage between circles ("I don't want Circle A to know I'm named Foo in Circle B")
  -   [ ] Migration to enforce no empty profile manes and no nominees without profiles (run 1 month after profile name is implemented.)
      - Remove `name` from nominees and Users tables
  - [x] Remove Functionality to edit profile name from members table
  - [ ] Only fetch profile name and not user.name
 
```
UPDATE profiles
SET name = users.name
FROM users
WHERE users.address IN (
  SELECT users.address 
  FROM users 
  GROUP BY users.address 
  HAVING COUNT(distinct LOWER(users.name)) = 1
) 
AND users.address = profiles.address
```

----

Note to Coordinape staff: see [this Notion page](https://www.notion.so/coordinape/2022-12-06-Dev-f11c01039cb14915ba61a1feed9ad0ef) for more context

### PR DESCRIPTION
**Pull request by sshmm**
## Motivation and Context

resolves #1395

## Description

1- Add a modal to ask users to enter a profile name that will be used in all circles
2- Change createusers mtations 'manual-address-entry' to check if the names exists in profiles and nominees
3- create users and profiles as follows:
 - if the user has a profile without a name the entered name will be the profile name
 - if the user doesn't have a profile, create a profile with that name
 - if the name exists for other addresses inform the user
 4- create Nominee only if the nominee name is not used by other nominees or users

### Test 1
1- Log in to coordinape
expected:
a modal should appear to enter user name and only accepts unique names

### Test 2
1- Create a nominee with a unique name
2- try to add user with that name
expected:
you should get an error about the situation

### Test 3
1- complete the vouching for the nominee in Test 2
expected: a profile with the nominee name should be added

### Test 4
1- create a user with a new name
expected: a profile with the user name should be added

### Test 5
1- create a user with a name in profiles table
expected: you should get an error

## Screenshots (if appropriate):
image

Screenshot from 2022-12-05 15-40-05

Screenshot from 2022-12-04 23-25-34

inform Admins of ignored names entry
image

### CODE REVIEW

File: api/hasura/actions/_handlers/createUsers.ts [hunk 1]
```diff
@@ -135,6 +137,162 @@ async function handler(req: VercelRequest, res: VercelResponse) {
...
+          names.length > 1
+            ? 'names used by other coordinape users'
+            : 'a name used by another coordinape user'
+        }: ${names}`,
+      },
+      422
+    );
+  }
+
+  const { profiles } = await adminClient.query(
```
Comment (zashton) [inline api/hasura/actions/_handlers/createUsers.ts:179]: You probably can save on the query call if you merge the create & update profile mutation into an upsert like this. Think this should work the same.

```
import { profiles_constraint, profiles_update_column } from '../../../../api-lib/gql/__generated__/zeus';

const createProfilesMutations = adminClient.mutate({
    insert_profiles:[
      {
        objects: newUsers.map( user => {
          return {
            address: user.address.toLowerCase(),
            name: user.name,
          }
        }),
        on_conflict: {
          constraint: profiles_constraint.profiles_address_key,
          update_columns: [profiles_update_column.name],
          where: {
            name: {_is_null: true}
          }
        }
      },
      {
        returning: {
          id: true,
        }
      }
    ]
  },
  {
    operationName: 'createUsers_createProfiles',
  });
  ```

File: api/hasura/actions/_handlers/createUserMutation.ts [hunk 2]
```diff
@@ -38,6 +42,63 @@ export async function createUserMutation(
...
+  if (!addressProfile && nameProfile) {
+    throw new UnprocessableError(
+      'This user name is used by another coordinape user'
+    );
+  }
+
+  let createProfileMutation = null;
+
+  if (!addressProfile && !nameProfile) {
+    createProfileMutation = await adminClient.mutate(
```
Comment (zashton) [inline api/hasura/actions/_handlers/createUserMutation.ts:59]: its possible to do it as an upsert as well here

File: src/lib/zod/index.ts [hunk 3]
```diff
@@ -122,6 +122,12 @@ export const generateApiKeyInputSchema = z
   })
   .strict();
 
+export const updateProfileNameSchemaInput = z
+  .object({
+    name: z.string(),
```
Comment (zashton) [inline src/lib/zod/index.ts:127]: should we impose a character limit ?

**Review by zashton — CHANGES_REQUESTED** [review]
nice work on the profile name changes, left a few suggestions.

File: api/hasura/actions/_handlers/createUsers.ts [hunk 4]
```diff
@@ -135,6 +137,162 @@ async function handler(req: VercelRequest, res: VercelResponse) {
...
+  //check if names are used by other coordinape users
+  const { profiles: existingNames } = await adminClient.query(
+    {
+      profiles: [
+        {
+          where: {
+            _or: newUsers.map(user => {
+              return {
+                _and: [
+                  { name: { _eq: user.name } },
```
[the remark this answers is not in the document]
Comment (sshmm) [reply api/hasura/actions/_handlers/createUsers.ts:149]: I am using citext for names so ilike is not necessary

**Review by levity — COMMENTED** [review]
Oops! I left a bunch of comments 2 days ago but never submitted the review. If anything I've said since then conflicts with a comment here, ignore the comment. :)

**Review by zashton — DISMISSED** [review]
Went through all the test cases, works with no issues. 👍

**Comment by zashton** [comment]
@sshmm note that all the e2e cypress tests are failing

**Review by zashton — DISMISSED** [review]

**Review by levity — APPROVED** [review]
nice work! sorry it took so long to get thru review 😅

**Pull request merged**